### PR TITLE
Allowing debug output to be written by telemetry processors

### PR DESCRIPTION
**Pull request by madeleinefletcher**

### CODE REVIEW

File: src/Core/Managed/Shared/Extensibility/Implementation/DebugOutput.cs [hunk 1]
```diff
@@ -8,11 +8,41 @@ namespace Microsoft.ApplicationInsights.Extensibility.Implementation
...
+    /// </summary>
+    public class DebugOutput : IDebugOutput
     {
-        public void WriteLine(string message)
+        /// <summary>
+        /// Write the specified <see cref="ITelemetry"/> item to debug output.
+        /// </summary>
+        /// <param name="telemetry">Item to write.</param>
+        /// <param name="filteredBy">If specified, indicates the telemetry item was filtered out and not sent to the API.</param>
+        public void WriteTelemetry(ITelemetry telemetry, string filteredBy = null)
```
Comment (madeleinefletcher) [inline src/Core/Managed/Shared/Extensibility/Implementation/DebugOutput.cs]: I'm open to suggestions on other places to put this method. It does need to be publicly exposed though, so the telemetry processors can access it.

File: src/TelemetryChannels/ServerTelemetryChannel/Shared/SamplingTelemetryProcessor.cs [hunk 2]
```diff
@@ -65,6 +66,7 @@ public void Process(ITelemetry item)
                         TelemetryChannelEventSource.Log.ItemSampledOut(item.ToString());
                     }
 
+                    new DebugOutput().WriteTelemetry(item, this.GetType().FullName);
```
Comment (madeleinefletcher) [inline src/TelemetryChannels/ServerTelemetryChannel/Shared/SamplingTelemetryProcessor.cs]: Should writing debug output be static?

File: src/Core/Managed/Shared/Extensibility/Implementation/DebugOutput.cs [hunk 3]
```diff
@@ -8,11 +8,41 @@ namespace Microsoft.ApplicationInsights.Extensibility.Implementation
...
+            {
+                string prefix = "Application Insights Telemetry: ";
+                if (string.IsNullOrEmpty(telemetry.Context.InstrumentationKey))
+                {
+                    prefix = "Application Insights Telemetry (unconfigured): ";
+                }
+
+                if (!string.IsNullOrEmpty(filteredBy))
+                {
+                    prefix = "Application Insights Telemetry (filtered by " + filteredBy + "): ";
```
Comment (madeleinefletcher) [inline src/Core/Managed/Shared/Extensibility/Implementation/DebugOutput.cs]: What do you think of this prefix? I can provide a regex so the debug output can be filtered for this.

**Comment by madeleinefletcher** [comment]
What do you think of this implementation?

File: src/Core/Managed/Shared/Extensibility/Implementation/TelemetryDebugWriter.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    using System;
+    using System.Diagnostics;
+    using Channel;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Platform;
+
+    /// <summary>
+    /// Writes telemetry items to debug output.
+    /// </summary>
+    public class TelemetryDebugWriter : IDebugOutput
```
Comment (SergeyKanzhelev) [inline src/Core/Managed/Shared/Extensibility/Implementation/TelemetryDebugWriter.cs]: You are right, we don't need interface anymore. Also make class static

File: src/TelemetryChannels/ServerTelemetryChannel/Shared/SamplingTelemetryProcessor.cs [hunk 5]
```diff
@@ -65,6 +66,7 @@ public void Process(ITelemetry item)
                         TelemetryChannelEventSource.Log.ItemSampledOut(item.ToString());
                     }
 
+                    TelemetryDebugWriter.WriteTelemetry(item, this.GetType().FullName);
```
Comment (SergeyKanzhelev) [inline src/TelemetryChannels/ServerTelemetryChannel/Shared/SamplingTelemetryProcessor.cs]: Does `this.GetType().FullName` includes namespaces? I'd say we only need name to make output more compact

**Comment by SergeyKanzhelev** [comment]
:shipit: with two comments

File: src/Core/Managed/Shared/Extensibility/Implementation/TelemetryDebugWriter.cs (hunk 4 above)
Comment (madeleinefletcher) [inline src/Core/Managed/Shared/Extensibility/Implementation/TelemetryDebugWriter.cs]: The are a lot of unit tests that actually test whether items are written to debug output. So I ended up reaching a compromise where the publicly exposed method is static, but we keep the interface for our unit tests.

File: src/TelemetryChannels/ServerTelemetryChannel/Shared/SamplingTelemetryProcessor.cs (hunk 5 above)
Comment (madeleinefletcher) [inline src/TelemetryChannels/ServerTelemetryChannel/Shared/SamplingTelemetryProcessor.cs]: Ok, makes sense.

File: src/Core/Managed/Shared/Extensibility/Implementation/TelemetryDebugWriter.cs (hunk 4 above)
Comment (SergeyKanzhelev) [inline src/Core/Managed/Shared/Extensibility/Implementation/TelemetryDebugWriter.cs]: ok. maybe time will come and we will clean it up =)

**Pull request merged**